autogen/coding/local_commandline_code_executor.py

```diff
@@ -0,0 +1,147 @@
+    2. When you need to perform some task with code, use the code to perform the task and output the result. Finish the task smartly.
+Solve the task step by step if you need to. If a plan is not provided, explain your plan first. Be clear which step uses code, and which step uses your language skill.
+When using code, you must indicate the script type in the code block. The user cannot provide any other feedback or perform any other action beyond executing the code you suggest. The user can't modify your code. So do not suggest incomplete code which requires users to modify. Don't use a code block if it's not intended to be executed by the user.
+If you want the user to save the code in a file before executing it, put # filename: <filename> inside the code block as the first line. Don't include multiple code blocks in one response. Do not ask users to copy and paste the result. Instead, use 'print' function for the output when relevant. Check the execution result returned by the user.
+"""
+
+        def add_to_agent(self, agent: LLMAgent) -> None:
+            """Add this capability to an agent."""
+            system_message = agent.system_message + self.DEFAULT_SYSTEM_MESSAGE_UPDATE
+            agent.update_system_message(system_message)
+
+    timeout: int = Field(default=60, ge=1, description="The timeout for code execution.")
+    work_dir: str = Field(default=".", description="The working directory for the code execution.")
+
+    @validator("work_dir")
+    def _check_work_dir(cls, v: str) -> str:
+        if os.path.exists(v):
+            return v
+        raise ValueError(f"Working directory {v} does not exist.")
+
+    @property
+    def user_capability(self) -> "LocalCommandlineCodeExecutor.UserCapability":
+        """Export a user capability that can be added to an agent."""
+        return LocalCommandlineCodeExecutor.UserCapability()
+
+    @property
+    def code_extractor(self) -> CodeExtractor:
+        """Export a code extractor that can be used by an agent."""
+        return MarkdownCodeExtractor()
+
+    def execute_code_blocks(self, code_blocks: List[CodeBlock]) -> CommandlineCodeResult:
+        """Execute the code blocks and return the result.
+
+        Args:
+            code_blocks (List[CodeBlock]): The code blocks to execute.
+
+        Returns:
+            CommandlineCodeResult: The result of the code execution."""
+        logs_all = ""
+        for i, code_block in enumerate(code_blocks):
+            lang, code = code_block.language, code_block.code
+            print(
+                colored(
+                    f"\n>>>>>>>> EXECUTING CODE BLOCK {i} (inferred language is {lang})...",
+                    "red",
+                ),
+                flush=True,
+            )
+            filename_uuid = uuid.uuid4().hex
+            filename = None
+            if lang in ["bash", "shell", "sh"]:
```

Comment:
Maybe there should be some security measure to prevent dangerous commands or keywords, for example:
```python
if "rm" in code:
    raise ValueError("Use of 'rm' command is not allowed.")
```

This would definitely be needed if someone was to opt to not user docker

Reply:
This could be a starting point: https://operavps.com/docs/dangerous-linux-commands/

Reply:
This is a good idea. The current executor doesn't do dangerous command removal when docker is turned off. Perhaps we can add these in the future? If we simply detect "rm" in the code it may be a lot of false positives.

Reply:
@AaronWard that's awesome!